File: Leetcode/210. Course Schedule II.py
class Solution:
    #using bfs i.e., Kahn's Algorithm to solve 
    def findOrder(self, n: int, p: List[List[int]]) -> List[int]:
        d=[0]*n
        adj=[[] for _ in range(n)]
        print(adj)
        for i in p:
            adj[i[1]].append(i[0])
            d[i[0]]+=1
        q = deque()
        for i in range(n):
            if d[i]==0:
                q.append(i)
        l=[]
        while q:
            t=q[0]
            q.popleft()
            l.append(t)
            for i in adj[t]:
                d[i]-=1
                if d[i]==0:
                    q.append(i)
        for i in d:
            if i!=0:
                return []
        return l
        

File: Leetcode/210. Course Schedule II.cpp
class Solution {
public:
    //Using DFS for now cpp implementation 
    //maintaining visited, deque, current path and if cycle is present then returning empty.
    bool dfs(int node,vector<int>& visited,vector<int>& cp,vector<vector<int>>& adj,deque<int>& d){
        visited[node]=1;
        cp[node]=1;
        for(int nbr:adj[node]){
            if(!visited[nbr]){
                if(dfs(nbr,visited,cp,adj,d)){
                    return true;
                }
            }else{
                if(cp[nbr]==1){
                    return true;
                }
            }
        }
        cp[node]=0;
        d.push_front(node);
        return false;

    }
    vector<int> findOrder(int n, vector<vector<int>>& prerequisites) {
        vector<int> visited(n,0),cp(n,0);
        vector<vector<int>> adj(n);
        deque<int> d;
        for(auto& i:prerequisites){
            adj[i[1]].push_back(i[0]);
        }
        for(int i=0;i<n;++i){
            if(!visited[i]){
                if(dfs(i,visited,cp,adj,d)){
                    return {};
                }
            }
        }
        vector<int> res;
        while(!d.empty()){
            res.push_back(d.front());
            d.pop_front();
        }
        return res;
    }
};
